components/tidb_query/src/codec/row/v2/encoder.rs

```diff
@@ -0,0 +1,295 @@
+}
+
+impl Row {
+    fn new(datums: Vec<Datum>, ids: &[i64]) -> Row {
+        let mut is_big = false;
+
+        let mut non_null_cols = Vec::with_capacity(ids.len());
+        let mut null_cols = Vec::with_capacity(ids.len());
+
+        let mut value_approx_size = 0;
+        for (datum, id) in datums.into_iter().zip(ids) {
+            let datum = flatten(datum);
+            if *id > 255 {
+                is_big = true;
+            }
+            value_approx_size += datum_size(&datum);
+
+            if datum == Datum::Null {
+                null_cols.push((datum, *id));
+            } else {
+                non_null_cols.push((datum, *id));
+            }
+        }
+        non_null_cols.sort_by_key(|(_, id)| *id);
+        null_cols.sort_by_key(|(_, id)| *id);
+
+        Row {
+            is_big,
+            non_null_cols,
+            null_cols,
+            value_approx_size,
+        }
+    }
+
+    // version(1) + flag(1) + columns_length(4) + ids_length + values_offset_length + values_length
+    #[inline]
+    fn approx_size(&self) -> usize {
+        6 + (self.non_null_cols.len() + self.null_cols.len()) * self.id_size()
+            + self.non_null_cols.len() * self.offset_size()
+            + self.value_approx_size
+    }
+
+    #[inline]
+    fn id_size(&self) -> usize {
+        if self.is_big {
+            LARGE_ID_SIZE
+        } else {
+            ID_SIZE
+        }
+    }
+
+    #[inline]
+    fn offset_size(&self) -> usize {
+        if self.is_big {
+            LARGE_OFFSET_SIZE
+        } else {
+            OFFSET_SIZE
+        }
+    }
+
+    #[inline]
+    fn offsets_len(&self) -> usize {
+        self.offset_size() * self.non_null_cols.len()
+    }
+}
+
+#[inline]
+fn datum_size(datum: &Datum) -> usize {
+    match *datum {
+        Datum::I64(i) => i64_size(i),
+        Datum::U64(u) => u64_size(u),
+        Datum::Bytes(ref bs) => bs.len(),
+        Datum::Dec(ref d) => d.approximate_encoded_size(),
+        Datum::Json(ref j) => j.binary_len(),
+        Datum::F64(_) | Datum::Time(_) | Datum::Dur(_) => {
+            unreachable!("these types should be flattened as numbers before")
+        }
+        _ => 0,
+    }
+}
+
+#[allow(clippy::match_overlapping_arm)]
+#[inline]
+fn i64_size(i: i64) -> usize {
+    match i {
+        i8::MIN_I64..=i8::MAX_I64 => 1,
+        i16::MIN_I64..=i16::MAX_I64 => 2,
+        i32::MIN_I64..=i32::MAX_I64 => 4,
+        _ => 8,
+    }
+}
+
+#[allow(clippy::match_overlapping_arm)]
+#[inline]
+fn u64_size(u: u64) -> usize {
+    match u {
+        u8::MIN_U64..=u8::MAX_U64 => 1,
+        u16::MIN_U64..=u16::MAX_U64 => 2,
+        u32::MIN_U64..=u32::MAX_U64 => 4,
+        _ => 8,
+    }
+}
+
+#[inline]
+fn flatten(datum: Datum) -> Datum {
+    match datum {
+        Datum::F64(f) => Datum::U64(f.to_bits()),
+        Datum::Time(t) => Datum::U64(t.to_packed_u64()),
+        Datum::Dur(d) => Datum::I64(d.to_nanos()),
+        _ => datum,
+    }
+}
```

Comment:
Prefer to remove this flatten function and embed its logic directly in the encoder, since it harms to the readability, i.e. making it not clear that how each data type is encoded in the encoder.

Reply:
done